src/Module.zig

```diff
@@ -386,6 +386,7 @@ pub const Decl = struct {
     /// the contents hash, followed by line, name, etc.
     /// For anonymous decls and also the root Decl for a File, this is 0.
     zir_decl_index: Zir.Inst.Index,
+    first_referenced_by: ReferencedBy = .{},
```

Comment:
`Decl` memory is something we are concerned with keeping minimal. Given that both of these things are true...

 * referenced-by traces are noise in some cases and it may make sense to opt-in to this extra diagnostic info
 * keeping track of referenced-by traces may come with a performance cost in the compiler (even when no compile errors occur)

...I suggest that we keep all referenced-by trace information in a separate table that remains empty unless the compilation opts into additional compile-error diagnostics.

Also, I would expect to see some kind of modification to `Module.destroyDecl`, removing `ReferencedBy` objects. Have you done any testing with incremental compilation on this? We may want to get some improved debug/testing infrastructure going for this before attempting to implement referenced by diagnostics.

Reply:
> Also, I would expect to see some kind of modification to `Module.destroyDecl`, removing `ReferencedBy` objects. Have you done any testing with incremental compilation on this?

I thought that removing the referencer would also make this decl unreferenced but that obviously only holds up if the decl is referenced from one place.